File: services/rust/actix/src/routes/handlers/put.rs
use crate::routes::{
    errors::AppError,
    extractors::{AppJson, AppPath},
};
use actix_web::{HttpResponse, ResponseError, web::Data};
use app::{state::AppState, task::Task};
use std::sync::Mutex;
use tracing::{error, info, instrument, warn};

#[instrument(skip_all)]
pub async fn put_handler(
    state: Data<Mutex<AppState>>,
    AppPath(id): AppPath,
    AppJson(new_task): AppJson<Task>,
) -> HttpResponse {
    if let Ok(mut state) = state.lock() {
        if let Some(task) = state.tasks.get_mut(&id) {
            let request = new_task;

            let previous_task = task.clone();
            *task = request.clone();

            info!(
                %id,
                from_secret = previous_task.secret.len(),
                to_secret = task.secret.len(),
                from_operation = previous_task.operation.to_string(),
                to_operation = task.operation.to_string(),
                method = "PUT",
                "Overwrote task"
            );

            return HttpResponse::Ok().json(task);
        } else {
            drop(state);
            warn!(%id, method = "PUT", "Task not found");
            return AppError::TaskNotFound.error_response();
        }
    }

    error!(%id, method = "PUT", "Poisoned lock");

    AppError::Internal.error_response()
}

File: services/go/echo/routes/handlers/put.go
package handlers

import (
	"app"
	"echo/routes"
	"log/slog"
	"net/http"
	"strings"

	"github.com/labstack/echo/v5"
)

func PutHandler(state *app.AppState) echo.HandlerFunc {
	return func(c *echo.Context) error {
		id, err := ParseID(c)
		if err != nil {
			return err
		}

		body, err := ReadBody(c)
		if err != nil {
			return err
		}

		newTask, err := ParseTask(c, body)
		if err != nil {
			return err
		}

		state.Mu.Lock()
		previousTask, ok := state.Tasks[id]
		if !ok {
			state.Mu.Unlock()
			slog.Warn("Task not found", "method", c.Request().Method, "path", c.Request().URL.Path, "id", id.String())
			return routes.AppErrors.TaskNotFound.Error(c)
		}
		state.Tasks[id] = newTask
		state.Mu.Unlock()

		slog.Info(
			"Overwrote task",
			"id", id.String(),
			"from_secret", len(previousTask.Secret),
			"to_secret", len(newTask.Secret),
			"from_operation", strings.ToLower(string(previousTask.Operation)),
			"to_operation", strings.ToLower(string(newTask.Operation)),
			"method", c.Request().Method,
		)

		return c.JSON(http.StatusOK, newTask)
	}
}
